docs/proposals/2018-05-declarative-delete-field.md

```diff
@@ -0,0 +1,86 @@
+# Declarative Delete Field
+
+Concept committed to [during 2018 April 12F2F](https://docs.google.com/document/d/1O7_fws7hwZ6qV3okAjbV5qdFaEq4bt1LYJ9bTu-IwvU)
+
+This document 2018-05.
+
+## Abstract
+
+Add a field to instance and binding to indicate that they have been asked to
+send a DELETE to the backend.
+
+## Motivation
+
+When the Originating-Identity header is used, the broker and refuse a delete
+call when the user info in the Originating-Identity header does not have
+permissions to delete the instance or binding. As this happens after kubernetes
+has accepted the DELETE as an true indication to do a delete, it has set the
+DeletionTimestamp, which is an irreversible step on the way to carrying out the
+necessary actions to delete a kubernetes resource.
+
+## Proposed Design
+
+ - Add a field to indicate deletion
+ - Add a condition to indicate permanent end state of deletion success.
+
+Once the field is activated the controller deprovisions the broker
+resource. When successfully deleted the controller then deletes the
+kubernetes resource.
+
+A feature flag to enable/disable this behavior.
+
+RBAC changes to allow the controller to issue deletes.
+
+Helm chart changes to enable the flag and the rbac rules.
+
+### API Resource Changes
+
+Add a condition for ServiceInstanceConditionType, ServiceBindingConditionType.
+```go
+  ServiceInstanceConditionDeleted ServiceInstanceConditionType = "Deleted"
+  ServiceBindingConditionDeleted ServiceBindingConditionType = "Deleted"
+```
+
+Add a field to spec of ServiceInstance, ServiceBinding.
+```go
+  ExistsExternally bool `json:"existsExternally,omitempty"`
+```
+
+This field is set to true by default. The user indicates that it wants
+to delete the backing resource by setting this field to false. The
+controller will see this update and attempt to delete the backing
+broker resource. If the broker delete succeeds, it sets the deleted
```

Comment:
The happy path is obvious, they work without this new flag (via normal Kubernetes deletion).
The non-happy paths are the ones that supposed to be improved with this change, thus before we cover them this flag is useless.

Reply:
The happy-path is the currently broken path.

Reply:
@mhbauer for example? Currently, if resource is successfully deprovisioned, we just remove the finalizer and proceed with deletion. Do I miss something?

Reply:
This isn't about successful vs uncsuccessful. 

This is about the delete not being allowed by the broker, yet kube will allow it. Once kube has allowed it, it cannot be unallowed. This denies the appropriate user the ability to do updates or deletes at all. The broker resource is stuck existing.